File: web-panel/src/lib/types/starlink.ts
export interface StarlinkStatus {
    available: boolean
    downlink_throughput_bps: number
    uplink_throughput_bps: number
    pop_ping_latency_ms: number
    pop_ping_drop_rate: number
    eth_speed_mbps: number
    obstruction_fraction: number
    currently_obstructed: boolean
    avg_prolonged_obstruction_duration_s: number
    avg_prolonged_obstruction_interval_s: number
    alert_thermal_throttle: boolean
    alert_thermal_shutdown: boolean
    alert_is_heating: boolean
    alert_slow_ethernet: boolean
    alert_power_save_idle: boolean
    alert_motors_stuck: boolean
    alert_no_ethernet_link: boolean
    alert_unexpected_location: boolean
    alert_roaming: boolean
    alert_mast_not_near_vertical: boolean
    alert_install_pending: boolean
    hardware_version: string
    software_version: string
    country_code: string
    uptime_s: number
    boot_count: number
    gps_valid: boolean
    gps_sats: number
    tilt_angle_deg: number
    boresight_azimuth_deg: number
    boresight_elevation_deg: number
    attitude_uncertainty_deg: number
    attitude_estimation_state: string
    desired_boresight_azimuth_deg: number
    desired_boresight_elevation_deg: number
    actuator_state: string
    has_actuators: string
    software_update_state: string
    software_update_progress: number
    outage_cause: string
    outage_duration_ns: number
    disablement_code: string
    mobility_class: string
    class_of_service: string
    cell_id: number
    satellite_id: number
    gateway_id: number
    on_backup_beam: boolean
    latitude: number
    longitude: number
    altitude: number
    is_snr_above_noise_floor: boolean
    is_snr_persistently_low: boolean
}

export interface StarlinkObstructionMap {
    num_rows: number
    num_cols: number
    snr: (number | null)[]
    max_theta_deg: number
    reference_frame: string
}

export interface StarlinkDataPoint {
    id: number
    node_id: number
    downlink_throughput_bps: number
    uplink_throughput_bps: number
    pop_ping_latency_ms: number
    pop_ping_drop_rate: number
    obstruction_fraction: number
    currently_obstructed: boolean
    tilt_angle_deg: number
    boresight_azimuth_deg: number
    boresight_elevation_deg: number
    gps_valid: boolean
    alert_flags: number
    created_at: string
}

export interface OutboundTestResult {
    success: boolean
    latency_ms: number
    status_code: number
    ip: string
    country: string
    error: string
    message: string
}

File: internal/agent/server/starlink.go
package server

import (
	"context"

	pb "github.com/nasnet-community/nasnet-panel-linux/pkg/agent/pb"
	"github.com/nasnet-community/nasnet-panel-linux/pkg/starlink"
)

// dishAddr resolves a request's dish address with the package default.
func dishAddr(req *pb.StarlinkRequest) string {
	if req != nil && req.DishAddress != "" {
		return req.DishAddress
	}
	return starlink.DefaultDishAddress
}

// GetStarlinkStatus queries the local Starlink dish and returns its status.
// On any failure (channel build, dial, RPC) we return Available=false instead
// of an RPC error so the panel can render a unified "Dish Unreachable" state
// without distinguishing transport vs application failure.
func (s *Server) GetStarlinkStatus(ctx context.Context, req *pb.StarlinkRequest) (*pb.StarlinkStatusResponse, error) {
	addr := dishAddr(req)

	client, err := starlink.SharedClient(addr)
	if err != nil {
		return &pb.StarlinkStatusResponse{Available: false}, nil
	}

	status, err := client.GetStatus(ctx)
	if err != nil {
		// Wedged channel — drop it so the next poll redials.
		starlink.EvictPooled(addr)
		return &pb.StarlinkStatusResponse{Available: false}, nil
	}

	return &pb.StarlinkStatusResponse{
		Available:                        status.Available,
		DownlinkThroughputBps:            float32(status.DownlinkThroughputBps),
		UplinkThroughputBps:              float32(status.UplinkThroughputBps),
		PopPingLatencyMs:                 float32(status.PopPingLatencyMs),
		PopPingDropRate:                  float32(status.PopPingDropRate),
		EthSpeedMbps:                     status.EthSpeedMbps,
		ObstructionFraction:              float32(status.ObstructionFraction),
		CurrentlyObstructed:              status.CurrentlyObstructed,
		AvgProlongedObstructionDurationS: float32(status.AvgProlongedObstructionDurationS),
		AvgProlongedObstructionIntervalS: float32(status.AvgProlongedObstructionIntervalS),
		AlertThermalThrottle:             status.AlertThermalThrottle,
		AlertThermalShutdown:             status.AlertThermalShutdown,
		AlertIsHeating:                   status.AlertIsHeating,
		AlertSlowEthernet:                status.AlertSlowEthernet,
		AlertPowerSaveIdle:               status.AlertPowerSaveIdle,
		AlertMotorsStuck:                 status.AlertMotorsStuck,
		AlertNoEthernetLink:              status.AlertNoEthernetLink,
		AlertUnexpectedLocation:          status.AlertUnexpectedLocation,
		AlertRoaming:                     status.AlertRoaming,
		AlertMastNotNearVertical:         status.AlertMastNotNearVert,
		AlertInstallPending:              status.AlertInstallPending,
		HardwareVersion:                  status.HardwareVersion,
		SoftwareVersion:                  status.SoftwareVersion,
		CountryCode:                      status.CountryCode,
		UptimeS:                          status.UptimeS,
		BootCount:                        status.BootCount,
		GpsValid:                         status.GpsValid,
		GpsSats:                          status.GpsSats,
		TiltAngleDeg:                     float32(status.TiltAngleDeg),
		BoresightAzimuthDeg:              float32(status.BoresightAzimuthDeg),
		BoresightElevationDeg:            float32(status.BoresightElevationDeg),
		SoftwareUpdateState:              status.SoftwareUpdateState,
		SoftwareUpdateProgress:           float32(status.SoftwareUpdateProgress),
		OutageCause:                      status.OutageCause,
		OutageDurationNs:                 status.OutageDurationNs,
		DisablementCode:                  status.DisablementCode,
		MobilityClass:                    status.MobilityClass,
		ClassOfService:                   status.ClassOfService,
		CellId:                           status.CellID,
		SatelliteId:                      status.SatelliteID,
		GatewayId:                        status.GatewayID,
		OnBackupBeam:                     status.OnBackupBeam,
		Latitude:                         status.Latitude,
		Longitude:                        status.Longitude,
		Altitude:                         status.Altitude,
		IsSnrAboveNoiseFloor:             status.IsSnrAboveNoiseFloor,
		IsSnrPersistentlyLow:             status.IsSnrPersistentlyLow,
		AttitudeUncertaintyDeg:           float32(status.AttitudeUncertaintyDeg),
		AttitudeEstimationState:          status.AttitudeEstimationState,
		DesiredBoresightAzimuthDeg:       float32(status.DesiredBoresightAzimuthDeg),
		DesiredBoresightElevationDeg:     float32(status.DesiredBoresightElevationDeg),
		ActuatorState:                    status.ActuatorState,
		HasActuators:                     status.HasActuators,
	}, nil
}

// GetStarlinkObstructionMap queries the local Starlink dish for its obstruction map.
// Mirrors the status path: on dial / RPC failure return an empty map (NumRows=0)
// so the panel can treat it identically to "dish unreachable" without a 500.
func (s *Server) GetStarlinkObstructionMap(ctx context.Context, req *pb.StarlinkRequest) (*pb.StarlinkObstructionMapResponse, error) {
	addr := dishAddr(req)

	client, err := starlink.SharedClient(addr)
	if err != nil {
		return &pb.StarlinkObstructionMapResponse{}, nil
	}

	m, err := client.GetObstructionMap(ctx)
	if err != nil {
		starlink.EvictPooled(addr)
		return &pb.StarlinkObstructionMapResponse{}, nil
	}

	return &pb.StarlinkObstructionMapResponse{
		NumRows:        m.NumRows,
		NumCols:        m.NumCols,
		Snr:            m.SNR,
		MaxThetaDeg:    m.MaxThetaDeg,
		ReferenceFrame: m.ReferenceFrame,
	}, nil
}
